(WIP): Dropdown

BeniCheni opened the pull request:
Refs #352 

Inspired by [Reach UI - MenuButton (Dropdown)](https://ui.reach.tech/menu-button) & its focus on accessibility, this PR builds on top of the components of Reach UI - MenuButton (Dropdown) to offer: `Dropdown`,`DropdownButton`, `DropdownList`, `DropdownItem`, `DropdownLink`, to support various dropdown compositions.


* Align the versions of deps among packages based on lerna bootstrap hoist warning. Should be warning free now when `npm run prepare` runs lerna bootstrap hoist.

TODO:
- [ ] Write doc
- [ ] Confirm code coverage
- [ ] Need a good way to include global styles, in order to include the base styles from `@reach/menu-button` via .css file.
- [ ] Update stories to improve the concept of Dropdown, versus the current stories still seem `Select`

P.S.: If this PR is successful as a foundation, the next phase of the Dropdown will expand to support composition such as mobile w. fullscreen (similar to the common Popover).

jxnblk commented on packages/dropdown/package.json:
```diff
@@ -0,0 +1,41 @@
+    ]
+  },
+  "peerDependencies": {
+    "pcln-design-system": ">=1.x || >=2.x",
```
This doesn't look like this is a dependency in the source code...

jxnblk commented:
According to the Reach docs, it looks like they only offer a `.css` file for including the base styles: https://ui.reach.tech/styling – and we currently don't have a standard way to bundle up CSS files in the design system. In order to use `@reach/menu-button` I think we'll need a good way to include global styles like this in our apps

BeniCheni commented:
@jxnblk,

> "peerDependencies": {
>    "pcln-design-system": ">=1.x || >=2.x",
> This doesn't look like this is a dependency in the source code...

Good catch! Removed.

>  In order to use @reach/menu-button I think we'll need a good way to include global styles ...

Added a TODO checkbox item in the 1st comment of this PR:
* Need a good way to include global styles, in order to include the base styles from @reach/menu-button via .css file.

Also per another convo about Dropdown stories in this PR being too much like `Select`, I've also added TODO checkbox item:
* Update stories to improve the concept of Dropdown, versus the current stories still seem Select

BeniCheni commented:
Posted an issue to ask about `style.css` in [#102 in `reach-ui`](https://github.com/reach/reach-ui/issues/102).

shakesBeardZ commented:
we need this, can some one review it

sdalonzo commented on packages/dropdown/package.json:
```diff
@@ -0,0 +1,40 @@
+  "description": "React component for dropdowns",
+  "main": "dist/index.js",
+  "scripts": {
+    "prepare": "babel src -d dist",
```
rollup please?

sdalonzo commented on packages/dropdown/package.json:
```diff
@@ -0,0 +1,40 @@
+    ]
+  },
+  "peerDependencies": {
+    "styled-components": ">=3.x || >= 4.x"
```
shouldn't react, react-dom and DS be peers as well?

sdalonzo commented on packages/dropdown/src/index.js:
```diff
@@ -0,0 +1,130 @@
+)
+
+const commonStyle = `
+  font-size: 12px;
```
use theme values?

sdalonzo commented on packages/dropdown/src/index.js:
```diff
@@ -0,0 +1,130 @@
+  text-decoration: none;
+`
+
+export { Dropdown, DropdownButton, DropdownList, DropdownItem, DropdownLink }
```
should these be attached instead, like Autocomplete and Stepper? ie `Dropdown.Button`, `Dropdown.List`, etc

sdalonzo commented on packages/dropdown/test/index.js:
```diff
@@ -0,0 +1,28 @@
+  DropdownLink
+} from '../src'
+
+describe('Dropdown', () => {
```
could you add some real tests, instead of just snaps?

BeniCheni commented:
@shakesBeardZ,

Thanks for your inquiry & interest in the `Dropdown` component. I'll champion a discussion w. the WG this coming session, and come with a path forward for this PR. I the related issue [#102 in reach-ui](https://github.com/reach/reach-ui/issues/102) was also closed w. a suggestion; pending to try it.

I'll update you again in this PR after the convo w. the WG. (cc @wesleymartin85 / @dansternfeld1 / @sdalonzo / @jes708 / @unscsprt )

Meanwhile, feel free to colla. to code review.

BeniCheni commented:
Unfortunately, after WG session, the feeling is that [`Select`](https://pricelinelabs.github.io/design-system/Select/) is serving most of what a `Dropdown` would offer. There was doubt that if reach-ui or other lib like downshift would be used. Also the underlying native `<select>` on mobile would opt in the mobile browser's UI & benefits for `<select>`.

On that note, this undesired `Dropdown` PR would be closed, as well as its related issue. Further questions could be discussed in the closed issue thought, #352 .
